rfc/011-registry-authentication.md

```diff
@@ -0,0 +1,281 @@
+
+1. Image pull secret referenced directly in the pod spec: <https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/>.
+2. Image pull secret attached to the service account: <https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#add-imagepullsecrets-to-a-service-account>. This indirectly populates the `imagePullSecrets` field of all pod specs using this service account.
+3. Global credentials in the container runtime configuration: <https://github.com/containerd/containerd/blob/5bcf77a/docs/cri/registry.md?plain=1#L37-L40>.
+4. Dedicated image credential providers: <https://kubernetes.io/docs/tasks/administer-cluster/kubelet-credential-provider/>.
+
+Image pull secrets come in [two formats](https://kubernetes.io/docs/concepts/configuration/secret/#docker-config-secrets).
+The `dockercfg` format has been [removed](https://github.com/docker/cli/pull/2504) by Docker itself, but this wasn't too long ago and it's likely still widely used.
+
+### Registry certificates
+
+Internal registries often use certificates that don't chain back to a web PKI root CA, but an internal one.
+Since Kubernetes doesn't have an API for specifying trusted registry CAs, such CA certificates need to be [configured at the container runtime](https://github.com/containerd/containerd/blob/5bcf77a55038ad658c57fdecc48af54935a0d42f/docs/cri/config.md?plain=1#L744).
+
+### Proxies
+
+Many corporate networks require HTTP connections to the public internet to go through a proxy.
+The de-facto standard for specifying such proxies are the environment variables `HTTP_PROXY`, `HTTPS_PROXY` and `NO_PROXY`.
+If Contrast is running in such an environment, it needs to respect these variables in order to pull images from within the guest.
+
+## Requirements
+
+1. Users must be able to guest-pull images from registries that require authentication.
+2. Users must be able to guest-pull images from registries that don't participate in web PKI.
+3. Users must be able to guest-pull images through HTTP proxies.
+
+## Design
+
+The overall idea of this design is to supply an image puller configuration to the VM that's _not measured_.
+Reason for this is that the image pull configuration is a workload owner secret that shouldn't be visible to (unauthenticated) verifiers - see [Alternatives considered](#alternatives-considered).
+
+### Changes to the `imagepuller`
+
+We add support for an optional configuration file to the `imagepuller`, with the following structure:
+
+```go
+import "github.com/google/go-containerregistry/pkg/authn"
+
+type InsecureConfig struct {
+    Registries map[string]Registry // mapping of domain name patterns to registry configuration
+    ExtraEnv map[string]string
+}
+
+type Registry struct {
+    authn.AuthConfig
+    CACerts string // concatenated, PEM-encoded CA certificates (system certs if nil)
+    InsecureSkipVerify bool
+}
```

Comment:
Using `string` for the certs is a bit inconvenient IMO. TOML spec allows the specification of an [array of tables](https://toml.io/en/v1.0.0#array-of-tables). How about just directly doing


```suggestion
type Registry struct {
    authn.AuthConfig
    CACerts []tls.Certificate
    InsecureSkipVerify bool
}
```

instead?